packages/gatsby/cache-dir/dev-loader.js

```diff
@@ -1,20 +1,123 @@
+
+          const cacheKey = msg.payload.id
+          const cachedStaticQueries = getStaticQueryResults()
+          const cachedResult = cachedStaticQueries[cacheKey]
+          if (didDataChange(newResult, cachedResult)) {
+            cachedStaticQueries[cacheKey] = newResult
+            ___emitter.emit(`staticQueryResult`, newResult)
+          }
+        } else if (msg.type === `pageQueryResult`) {
+          const newPageData = msg.payload.result
+
+          const pageDataDbCacheKey = normalizePagePath(msg.payload.id)
+          const cachedPageDataDb = getPageDataDb()
+          const cachedPageData = cachedPageDataDb.get(pageDataDbCacheKey)
+            ?.payload
+
+          if (didDataChange(newPageData, cachedPageData)) {
+            // always update canonical key for pageDataDb
+            cachedPageDataDb.set(pageDataDbCacheKey, {
+              pagePath: pageDataDbCacheKey,
+              payload: newPageData,
+              status: `success`,
+            })
+
+            const cachedPageDb = getPageDb()
+            const cachedPage = cachedPageDb.get(pageDataDbCacheKey)
+            if (cachedPage) {
+              cachedPageDb.set(
+                pageDataDbCacheKey,
+                mergePageEntry(cachedPage, newPageData)
+              )
+            }
+
+            // Additionally if those are query results for "/404.html"
+            // we have to update all paths user wanted to visit, but didn't have
+            // page for it, because we do store them under (normalized) path
+            // user wanted to visit
+            if (pageDataDbCacheKey === `/404.html`) {
+              this.notFoundPagePathsInCaches.forEach(notFoundPath => {
+                const previousPageDataEntry = cachedPageDataDb.get(notFoundPath)
+                if (previousPageDataEntry) {
+                  cachedPageDataDb.set(notFoundPath, {
+                    ...previousPageDataEntry,
+                    payload: newPageData,
+                  })
+                }
+
+                const previousPageEntry = cachedPageDb.get(notFoundPath)
+                if (previousPageEntry) {
+                  cachedPageDb.set(
+                    notFoundPath,
+                    mergePageEntry(previousPageEntry, newPageData)
+                  )
+                }
+              })
+            }
+
+            ___emitter.emit(`pageQueryResult`, newPageData)
+          }
+        }
+      })
+    } else {
+      console.warn(`Could not get web socket`)
+    }
   }
 
   loadPage(pagePath) {
     const realPath = findPath(pagePath)
-    return super.loadPage(realPath).then(result =>
-      require(`./socketIo`)
-        .getPageData(realPath)
-        .then(() => result)
-    )
+    return super.loadPage(realPath).then(result => {
+      const isNotFound = this.pageDb.get(realPath)?.notFound
+
+      if (isNotFound) {
+        this.notFoundPagePathsInCaches.add(realPath)
+      }
+
+      return result
+    })
```

Comment:
No big deal but just pointing out that we have `isPageNotFound`
```suggestion
    return super.loadPage(realPath).then(result => {
      if (this.isPageNotFound(realPath)) {
        this.notFoundPagePathsInCaches.add(realPath)
      }

      return result
    })
```